src/embeddings/embeddings_calculator.cc

```diff
@@ -0,0 +1,270 @@
+                if (it->value.IsString()) {
+                    response += it->value.GetString();
+                } else {
+                    return absl::InvalidArgumentError("user should be string");
+                }
+            }
+        } else {
+            return absl::InvalidArgumentError("Input is empty");
+        }
+
+        // Automatically deduce tokenizer input name
+        std::vector<std::string> tokenizerInputNames = tokenizer_session->getInputNames();
+        std::vector<std::string> embeddingsInputNames = embeddings_session->getInputNames();
+        RET_CHECK(tokenizerInputNames.size() == 1);
+        const std::string& tokenizerInputName = tokenizerInputNames.at(0);
+        LOG(INFO) << "Tokenizer input name detected: " << tokenizerInputName;
+
+        ::InferenceInput tokenizerInputMap;
+        tokenizerInputMap[tokenizerInputName] = ov::Tensor{
+            ov::element::string,
+            ov::Shape{input_strings.size()},
+            input_strings.data()};
+
+        ::InferenceOutput embeddingsOutputMap;
+        try {
+            ::InferenceOutput tokenizerOutputMap = tokenizer_session->infer(tokenizerInputMap);
+            ::InferenceInput embeddingsInputMap;
+            RET_CHECK(tokenizerOutputMap.size() >= embeddingsInputNames.size());
+            for (const auto& embeddingsInputName : embeddingsInputNames) {
+                auto it = tokenizerOutputMap.find(embeddingsInputName);
+                RET_CHECK(it != tokenizerOutputMap.end());
+                embeddingsInputMap[embeddingsInputName] = it->second;
+            }
+            embeddingsOutputMap = embeddings_session->infer(embeddingsInputMap);
+            RET_CHECK(embeddingsOutputMap.size() > 0);
+        } catch (const std::exception& e) {
+            LOG(INFO) << "Caught exception from session infer():" << e.what();
+            RET_CHECK(false);
+        } catch (...) {
+            LOG(INFO) << "Caught unknown exception from session infer()";
+            RET_CHECK(false);
+        }
+
+        // TODO (bstrzele) (dtrawins): Support both kinds of models with 1 or 2 outputs?
+        // TODO: At the moment hardcoded for 1 output only supporting bge-large-en-v1.5, but will not work with rerank/other models
+        // TODO: Must require last_hidden_state kind of tensor
+
+        ov::Tensor lastHiddenStateTensor; 
+        if (embeddingsOutputMap.size() == 2) { // GTE
+            // Search by number of dimensions, should be 3
+            bool found = false;
+            for (const auto& [name, tensor] : embeddingsOutputMap) {
+                if (tensor.get_shape().size() == 3) {
+                    lastHiddenStateTensor = tensor;
+                    found = true;
+                    break;
+                } 
+            }
+            RET_CHECK(found);
+        } else { // BGE
+            RET_CHECK(embeddingsOutputMap.size() == 1);
+            lastHiddenStateTensor = embeddingsOutputMap.begin()->second;
```

Comment:
maybe we should call it embeddingsTensor?

Reply:
done